tests/test_internals.cc

```diff
@@ -0,0 +1,92 @@
+/**
+ * \file   test_internals.cc
+ * \author Lars Froehlich
+ * \date   Created on November 4, 2022
+ * \brief  Test suite for internal functions.
+ *
+ * \copyright Copyright 2022 Deutsches Elektronen-Synchrotron (DESY), Hamburg
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation, either version 2.1 of the license, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+// SPDX-License-Identifier: LGPL-2.1-or-later
+
+#include <gul14/cat.h>
+#include <gul14/catch.h>
+
+#include "../src/internals.h"
+
+using gul14::cat;
+using namespace task;
+using namespace Catch::Matchers;
+
+TEST_CASE("remove_abort_markers_from_error_message()", "[internals]")
+{
+    std::string msg;
+
+    SECTION("Empty message")
+    {
+        REQUIRE(remove_abort_markers_from_error_message(msg) == ErrorCause::uncaught_error);
+        REQUIRE(msg.empty());
+    }
+
+    SECTION("Normal error message")
+    {
+        msg = "This is an error message";
+        REQUIRE(remove_abort_markers_from_error_message(msg) == ErrorCause::uncaught_error);
+        REQUIRE(msg == "This is an error message");
+    }
+
+    SECTION("Script termination (abort marker(s) without message)")
+    {
+        msg = std::string(abort_marker);
```

Comment:
We require only one `abort_marker` as a `terminated_by_script` error code? Why do we have this case? In the code at least I set two `abort_marker`'s around an empty string.

Reply:
> Just a small question (not about this PR but the feature in general) Why do we signal the abort in-band (i.e. in the message text) and not out-of-band (oob)? This feels like a quickly added feature and 'we add this kludge now, but will implement it properly later'.

This in-band signaling right now fulfills two functions:
1. Signal the intent to abort the sequence from within Lua code or C++ code within Lua (abort can be caused by explicit user request, timeout, explicit call to terminate_sequence())
2. Delimit the "payload" error message – for instance, if we generate a Lua error with the message "EEE", Lua and sol will add a prefix and a suffix to it and the resulting message will be `sol: runtime error: [string "error("EEE"):1: EEE stack traceback: [C]: in function 'main' [string "error("EEE"..."]:1: in main chunk` or something along those lines. By adding the abort marker to the left and to the right of the "EEE", we can identify it easily.  

> I'm maybe not properly 'in' the code, and I have a hard time finding the places where it is triggered. But is not always a lua::state also available in that places? Why not store a `bool` in the state, that says `we are totally messed up and can not be catched`?

I guess that could be done as well, we should always have access to the Lua state. This error state should be picked up by `Step::execute()` and could then be packaged into an `AbortException` or something like this.

In any case, that would be a major change that I would rather not address in this PR. Even if we do not like the in-band abort signaling that we have now, this PR at least extracts the functionality to deal with it and puts it to the side, where we can also remove it more easily. More importantly, it fixes the very ugly reporting of `Step` errors that we still generate:
```
ror: sol: runtime error: [string "dset("XFEL_SIM.MAGNETS/MAGNET.ML/BB.62.I1D/KI..."]:8: 🛑ABORT🛑Stop on user request🛑ABORT🛑 stack traceback: [C]: in function 'base.sleep' [string "dset("XFEL_SIM.MAGNETS/MAGNET.ML/BB.62.I1D/KI..."]:8: in main chunk
```

Reply:
Yes, the Q was in general, not for this concrete PR. As this in band signalling gets more and more refined the switch to oob becomes less and less likely ;) That's why I asked.